### PR TITLE
fix(gitpod-cli): ports list safe access to exposed url

### LINKED ISSUE #11990: Segmentation fault when running "gp ports list" in a task command
### Bug description

I started a workspace with two simple tasks :

```yaml
tasks:
  # Console for the user with a welcome message
  - name: Console
    command: |
      sleep 10
      echo "========================================"
      echo "Welcome to this repository ! Hugo server should start automatically and show both drafts and future posts, see below for the URL."
      gp ports list
      echo "========================================"
  # Dedicated terminal to display the logs
  - name: Logs
    command: |
      hugo serve --buildDrafts --buildFuture
    openMode: split-right

```

When opening gitpod I see a segmentation fault in the first console, when running "gp ports list" :

```bash
HISTFILE=/workspace/.gitpod/cmd-0 history -r; {
> echo "========================================"
> echo "Welcome to this repository ! Hugo server should start automatically and show both drafts and future posts, see below for the URL."
> gp ports list
> echo "========================================"
> 
> }
========================================
Welcome to this repository ! Hugo server should start automatically and show both drafts and future posts, see below for the URL.
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x95560e]

goroutine 1 [running]:
github.com/gitpod-io/gitpod/gitpod-cli/cmd.glob..func11(0xf70c80, {0xa9a3e2, 0x0, 0x0})
        /tmp/build/components-gitpod-cli--app.6b8796aa57fed8a61e36573e2558f89f4aa872c2/cmd/ports-list.go:86 +0x3ae
github.com/spf13/cobra.(*Command).execute(0xf70c80, {0xfb1318, 0x0, 0x0})
        /home/gitpod/go-packages/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:856 +0x663
github.com/spf13/cobra.(*Command).ExecuteC(0xf70780)
        /home/gitpod/go-packages/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:960 +0x39d
github.com/spf13/cobra.(*Command).Execute(...)
        /home/gitpod/go-packages/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:897
github.com/gitpod-io/gitpod/gitpod-cli/cmd.Execute()
        /tmp/build/components-gitpod-cli--app.6b8796aa57fed8a61e36573e2558f89f4aa872c2/cmd/root.go:35 +0x325
main.main()
        /tmp/build/components-gitpod-cli--app.6b8796aa57fed8a61e36573e2558f89f4aa872c2/main.go:12 +0x17
========================================
```

Trying to run the command manually after the workspace as fully setup is OK : 
```bash
gitpod /workspace/information-platform (master) $ gp ports list
| PORT |     STATUS     |                                URL                                 | NAME & DESCRIPTION |
|------|----------------|--------------------------------------------------------------------|--------------------|
| 1313 | open (private) | https://1313-[REDACTED].ws-eu59.gitpod.io | hugo-server        |
```

### Steps to reproduce

My .gitpod.io file is very basic. 

```yaml
tasks:
  # Console for the user with a welcome message
  - name: Console
    command: |
      sleep 10
      echo "========================================"
      echo "Welcome to this repository ! Hugo server should start automatically and show both drafts and future posts, see below for the URL."
      gp ports list
      echo "========================================"
  # Dedicated terminal to display the logs
  - name: Logs
    command: |
      hugo serve --buildDrafts --buildFuture
    openMode: split-right

# List the ports to expose. Learn more https://www.gitpod.io/docs/config-ports/
ports:
  - port: 1313
    name: hugo-server
    onOpen: open-browser

```

### Workspace affected

_No response_

### Expected behavior

The gp ports list command should display the open ports

### Example repository

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by andreafalzetti**
## Description
This PR fix safe access to `port.Exposed.*` for both `Url` and `Visibility`.

Also, I've refactored the existing logic to determine the status, to be more readable and to follow the [vscode implementation](https://github.com/gitpod-io/openvscode-server/blob/gp-code/main/extensions/gitpod-shared/src/workspacePort.ts#L116-L140)

```
 gp ports list
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x97c5ae]

goroutine 1 [running]:
github.com/gitpod-io/gitpod/gitpod-cli/cmd.glob..func12(0xfb5fe0, {0xac9654, 0x0, 0x0})
        github.com/gitpod-io/gitpod/gitpod-cli/cmd/ports-list.go:86 +0x3ae
github.com/spf13/cobra.(*Command).execute(0xfb5fe0, {0xff6aa8, 0x0, 0x0})
        github.com/spf13/cobra@v1.1.3/command.go:856 +0x663
github.com/spf13/cobra.(*Command).ExecuteC(0xfb5ae0)
        github.com/spf13/cobra@v1.1.3/command.go:960 +0x39d
github.com/spf13/cobra.(*Command).Execute(...)
        github.com/spf13/cobra@v1.1.3/command.go:897
github.com/gitpod-io/gitpod/gitpod-cli/cmd.Execute()
        github.com/gitpod-io/gitpod/gitpod-cli/cmd/root.go:35 +0x325
main.main()
        github.com/gitpod-io/gitpod/gitpod-cli/main.go:12 +0x17
```

## Related Issue(s)

Fixes #11990

## How to test
- Start a workspace in prev env: https://afalz-fix-962bf6477c.preview.gitpod-dev.com/#https://github.com/gitpod-io/empty
- Start a workspace in production from this PR https://gitpod.io/#https://github.com/gitpod-io/gitpod/pull/13680
- in the prod workspace, check `kubectx` to make sure you are in the context of this prev env, otherwise run `previewctl install-context`
- in the prod workspace, scale down servers `kubectl scale --replicas=0 deployment server` 
- in the prev env workspace, try to trigger the port exposure with `curl lama.sh | sh -s -- -p 3333`
- in the prev env workspace, run `gp ports list` and make sure it does not throw any error
- observe how `gp ports list` shows missing exposed URL instead of crashing

## Release Notes

```release-note
Bugfix for gitpod CLI: avoid failure on `ports list` when port is not exposed
```

## Werft options:

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`

### CODE REVIEW

**Comment by roboquat** [comment]
@andreafalzetti: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by andreafalzetti** [comment]
@akosyakov I've updated and followed the steps with @felladrin, however even by shutting down the server, somehow we were able to expose the port. Is there any cmd/step missing?

**Comment by andreafalzetti** [comment]
I've updated the how to test sectioc.

@felladrin could you help reviewing? 🙏

File: components/gitpod-cli/cmd/ports-list.go [hunk 1]
```diff
@@ -59,7 +64,7 @@ var listPortsCmd = &cobra.Command{
 					status = "detecting..."
 					statusColor = tablewriter.FgYellowColor
 				}
-			} else if port.Served {
+			} else if port.Served && port.Exposed != nil {
```
Comment (akosyakov) [inline components/gitpod-cli/cmd/ports-list.go:67]: Should not it be `served` if exposed is nil?

Comment (andreafalzetti) [reply components/gitpod-cli/cmd/ports-list.go]: I've refactored the code

**Comment by andreafalzetti** [comment]
> @andreafalzetti Good to test?

The preview environment seems to be broken even if the build is green, so I wasn't able to test myself and open for review yet :(

**Comment by andreafalzetti** [comment]
Closed in favour of https://github.com/gitpod-io/gitpod/pull/14040/files because the prev env is in an unrecoverable state. cc @meysholdt

**Pull request closed without merging**